plugins/lsp/conform-nvim.nix

```diff
@@ -36,97 +86,159 @@ in
+
+      notify_no_formatters = defaultNullOpts.mkBool true "Conform will notify you when no formatters are available for the buffer.";
+
+      formatters =
+        defaultNullOpts.mkAttrsOf types.anything { }
+          "Custom formatters and changes to built-in formatters.";
     };
 
-    formatAfterSave = helpers.defaultNullOpts.mkNullable' {
-      type =
-        with helpers.nixvimTypes;
-        either strLuaFn (submodule {
-          options = {
-            lspFallback = mkOption {
-              type = types.bool;
-              default = true;
-              description = "See :help conform.format for details.";
-            };
-          };
-        });
-      # Unknown plugin default
-      description = ''
-        If this is set, Conform will run the formatter asynchronously after save.
-        It will pass the table to conform.format().
-        This can also be a function that returns the table.
-      '';
+  settingsExample = {
+    formattersByFt = {
+      bash = [
+        "shellcheck"
+        "shellharden"
+        "shfmt"
+      ];
+      cpp = [ "clang_format" ];
+      javascript = {
+        __unkeyed-1 = "prettierd";
+        __unkeyed-2 = "prettier";
+        timeout_ms = 2000;
+        stop_after_first = true;
+      };
+      "_" = [
+        "squeeze_blanks"
+        "trim_whitespace"
+        "trim_newlines"
+      ];
     };
+    format_on_save = # Lua
+      ''
+        function(bufnr)
+          if vim.g.disable_autoformat or vim.b[bufnr].disable_autoformat then
+            return
+          end
 
-    logLevel = helpers.defaultNullOpts.mkLogLevel "error" ''
-      Set the log level. Use `:ConformInfo` to see the location of the log file.
-    '';
+          if slow_format_filetypes[vim.bo[bufnr].filetype] then
+            return
+          end
 
-    notifyOnError = helpers.defaultNullOpts.mkBool true "Conform will notify you when a formatter errors";
+          local function on_format(err)
+            if err and err:match("timeout$") then
+              slow_format_filetypes[vim.bo[bufnr].filetype] = true
+            end
+          end
 
-    formatters = helpers.defaultNullOpts.mkAttrsOf' {
-      type = types.anything;
-      # Unknown plugin default
-      description = "Custom formatters and changes to built-in formatters";
-    };
-  };
+          return { timeout_ms = 200, lsp_fallback = true }, on_format
+         end
+      '';
+    format_after_save = # Lua
+      ''
+        function(bufnr)
+          if vim.g.disable_autoformat or vim.b[bufnr].disable_autoformat then
+            return
+          end
 
-  config =
-    let
-      setupOptions =
-        with cfg;
-        {
-          formatters_by_ft = formattersByFt;
-          format_on_save =
-            if builtins.isAttrs formatOnSave then
-              {
-                lsp_fallback = formatOnSave.lspFallback;
-                timeout_ms = formatOnSave.timeoutMs;
```

Comment:
I'm not sure users who had `formatOnSave.*` defined in their configs will get a graceful migration to the new snake_case names in `settings.format_on_save.*`... Seeing as it's a union type, I can't think of a robust way to offer such migration either... :thinking:

Reply:
Not sure what you mean by this? It's similar to other migrations from the renamedToSettings we've been doing.